compiler/src/dotty/tools/dotc/core/TypeErasure.scala

```diff
@@ -556,13 +585,19 @@ import TypeErasure._
  *                        If false, they are erased like normal classes.
  *  @param isConstructor  Argument forms part of the type of a constructor
  *  @param isSymbol       If true, the type being erased is the info of a symbol.
- *  @param wildcardOK     Wildcards are acceptable (true when using the erasure
- *                        for computing a signature name).
+ *  @param inSigName      This eraser is used for `TypeErasure.sigName`,
+ *                        see `TypeErasure#apply` for more information.
  */
-class TypeErasure(sourceLanguage: SourceLanguage, semiEraseVCs: Boolean, isConstructor: Boolean, isSymbol: Boolean, wildcardOK: Boolean) {
+class TypeErasure(sourceLanguage: SourceLanguage, semiEraseVCs: Boolean, isConstructor: Boolean, isSymbol: Boolean, inSigName: Boolean) {
 
-  /**  The erasure |T| of a type T. This is:
+  /**  The erasure |T| of a type T. 
+   *
+   *   If computing the erasure of T requires erasing a WildcardType or an
+   *   uninstantiated type variable, then an exception signaling an internal
+   *   error will be thrown, unless `inSigName` is set in which case `null`
+   *   will be returned.
```

Comment:
Why is `null` used in this case? We don't usually use `null` to signal error conditions except in a local context where it means "variable not yet defined". I think we will get a more streamlined design if we keep `WildcardType` as a possible result of erasure and map uninstantiated variables to `WildcardType`. `WildcardType`s already map to `tpnme.uninstantiated` in `sigName`, so no special logic is needed. 

All we need is to propagate `WildcardType`s outwards in `Erasure.apply`. The most elegant way to do this 
would be to handle them in `erasedLub` and `erasedGlb`.

Reply:
> I think we will get a more streamlined design if we keep WildcardType as a possible result of erasure and map uninstantiated variables to WildcardType.  [...] All we need is to propagate WildcardTypes outwards in Erasure.apply

I chose to use `null` so that the compiler would ensure that I don't forget to propagate the special value outwards, I think type erasure is both tricky and critical enough that it's worth encoding more invariants in the type system if we can. I can switch to WildcardType if you prefer but I don't think the end result will be shorter.

Reply:
Let's try with `WildcardType`. I think propagation is quite localized here, so it's not that hard to verify that it indeed happens.